### ISSUE TITLE
About type changes in backward pass

### ISSUE THREAD

**Issue by GiggleLiu** [under-Peter/OMEinsum.jl#16]
Since `t` is real, as defined in
https://github.com/under-Peter/OMEinsum.jl/blob/426907186a16bfaf9761009bc50c45d961f908ad/test/autodiff.jl#L63

Please check if the gradient is converted to `real` correctly.
With the `mutate` branch of `Zygote`, tests in line 90-91 fails.

**Comment by under-Peter** [comment]
MWE for the problem is generally:
```
julia> using Zygote

julia> a, b = rand(2,2), rand(ComplexF64,2,2);

julia> Zygote.gradient(sum∘*,a,b)
```
where the gradient for both `a` and `b` will be complex. 

The alternatives are:
1. return the complex gradient, accepting that e.g. gradient descent is not type stable
2. return some projection to the reals, keeping type stability but possibly losing a lot of information

If we use 1., we put responsibility on the user to pick the correct types for their inputs, e.g. converting an initially real matrix to a complex, which is somewhat reasonable but might lead to subtle errors which are hard for us to decide.

If we use 2., we would require users to use e.g. `hook`s ([as described here](https://fluxml.ai/Zygote.jl/dev/adjoints/#Gradient-Hooks-1)) to convert the complex gradient to a real gradient. But we would also always return the true gradient. 

I tend to option two, putting the responsibility on the user if they *don't* want the complex gradient, what do you think @GiggleLiu ?

**Comment by GiggleLiu** [comment]
Yes, option 2 is better.

**Comment by under-Peter** [comment]
Sorry, I switched 1. and 2. in the text. Which one do you mean, keeping complex or projecting to reals?

**Comment by GiggleLiu** [comment]
I think not changing type is better.

**Comment by GiggleLiu** [comment]
Closed by #56
